---
title: "Interpreting *de novo* variation in human disease using **denovolyzeR**"
author: "Authors: James S. Ware^1-4^*, Kaitlin E. Samocha^1-3^, Jason Homsy^1,5^, Mark J. Daly^1-3^"
output:
  word_document:
    reference_docx: template.docx
---

Alternative title ideas:  
A statistical approach to assess the frequency of *de novo* variation  
Interpreting *de novo* variation in human disease: a model-based approach  

Contact information  
^1^ Department of Genetics, Harvard Medical School, Boston MA  
^2^ Broad Institute of MIT and Harvard, Cambridge MA  
^3^ Analytical and Translational Genetics Unit, Massachusetts General Hospital and Harvard Medical School, Boston MA  
^4^ NIHR Cardiovascular Biomedical Research Unit at Royal Brompton Hospital and Imperial College London, London UK  
^5^ Cardiovascular Research Center, Massachusetts General Hospital, Boston MA
* corresponding author

###This will be removed from final manuscript, but prevents permutations from re-running every time manuscript is updated during draft phase...
```{r}
library(knitr)
opts_chunk$set(cache=T)
```

##ABSTRACT
Spontaneously arising (*de novo*) genetic variants are important in human disease, yet every individual carries many such variants, with a median of 1 *de novo* variant affecting the protein-coding portion of the genome.  A recently described mutational model {25086666} provides a powerful framework for the robust statistical evaluation of such coding variants, enabling the interpretation of *de novo* variation in human disease.  Here we describe a new open-source software package, **denovolyzeR**, that implements this model and provides tools for the analysis of *de novo* coding sequence variants.

###Keywords: *de novo* variant, exome sequencing

##INTRODUCTION

Spontaneously arising (*de novo*) genetic variants are important in human disease. Every individual carries approximately 100 such variants that have arisen via mutational events in the parental germline or early embryo, with a median of 1 *de novo* variant affecting the protein-coding portion of genome, referred to as the exome {21666693; 20080596}.

Exome sequencing and analysis of *de novo* variants has successfully identified genes underlying rare and genetically homogeneous Mendelian diseases. In Kabuki syndrome, for example, non-synonymous *de novo* variants were identified in KMT2D (MLL2) in 9 out of 10 unrelated individuals {20711175}. An accumulation of this magnitude would be extremely improbable in the absence of a causal role in the disease given both the rarity and independence of *de novo* variants.

By contrast, it is more challenging to dissect the role of *de novo* variants in conditions with high levels of locus heterogeneity, including heritable complex traits and some Mendelian conditions, where *de novo* variants may be spread across many genes, and may make a smaller overall contribution to pathogenesis. Here it may be possible to assess the global contribution of *de novo* coding variants to disease by comparing their frequency in cases and controls, given sufficiently large sample sizes. However, at the level of individual genes, the interpretation of *de novo* variants is complicated by the background mutation rate, which varies greatly between genes. Additionally, as more individuals are sequenced, it is inevitable that multiple *de novo* variants will be observed in some genes by chance.

A statistical framework has recently been developed to address these challenges {25086666}, with respect to *de novo* single nucleotide variants (SNVs) in coding sequence. Briefly, the mutability of each gene is individually determined based on local sequence context, and the probability that a *de novo* event will arise in a single copy of the gene in one generation is calculated. The consequence of each possible *de novo* SNV is computed, and *de novo* probabilities are tabulated for each variant class. In order to more fully assess loss-of-function (lof) variation, the probability of a frameshifting insertion or deletion is also estimated for all genes (proportional to the length of the gene and the ratio of nonsense to frameshifting indels genome-wide under the assumption that the two classes have similar selective pressure against them). For a given study population, *de novo* variants can be evaluated by comparing the observed numbers of variants with the number expected based on this model and the population size, using a Poisson framework. This permits robust significance estimates for the pileup of *de novo* variation in individual genes and gene sets, and increases the power of genome-wide analyzes.

In this unit, we describe the application of this statistical framework to analyze *de novo* variants using **denovolyzeR** {http://denovolyzer.org}, an open-source software package written for the `R` statistical software environment {R: A Language and Environment for Statistical Computing}. We present protocols for four analyzes: to assess (i) whether there is a genome-wide excess of *de novo* variation for different functional classes of variant, (ii) whether there is a genome-wide excess of genes with multiple *de novo* variants, (iii) whether individual genes carry an excess of *de novo* variants, and (iv) whether a pre-specified set of genes collectively shows an enrichment of *de novo* variants. 


#BASIC PROTOCOL 1
##Assessing the genome-wide burden of *de novo* variants
This protocol will assess whether there is a genome-wide excess of *de novo* variation for different functional classes of variant.

###Materials

- A computer running the `R` software environment, available for UNIX platforms, Windows and MacOS from http://www.r-project.org.  
- The **denovolyzeR** package. The latest stable release can be installed directly from the Comprehensive R Archive Network (CRAN) from within `R`

```
install.packages("denovolyzeR")
```
- Other download and installation options, including for the latest development version, are described at http://denovolyzer.org  
- `dplyr` and `reshape` packages.  These dependencies may be installed automatically when **denovolyzeR** is installed (depending on your installation route). Otherwise they can be installed by running:

```
install.packages("dplyr","reshape")
```

- A table of *de novo* variants.  The minimum input comprises two columns of data: gene names, and variant classes (functional consequence of each variant).  
Example data is included in the **denovolyzeR** package, and will be used in this protocol. The dataset comprises a `data.frame` of *de novo* variants identified in 1078 individuals with autism {25086666}, named `autismDeNovos`.  
It is assumed that readers are able to import their own data into the `R` environment, using the `read.table` function or equivalent (in R, `?read.table` will provide help).  

1) In R, load the **denovolyzeR** package.
```{r, message=FALSE, warning=FALSE}
library(denovolyzeR)
```

2) Prepare input data. 
View demonstration data provided with the **denovolyzeR** package.  Alternatively, users may import their own data in an equivalent format.
```{r}
dim(autismDeNovos); head(autismDeNovos)
```

Variant classes must be labeled using the following terms:  
"syn" (synonymous), "mis" (missense), "non" (nonsense), "splice" (canonical splice site) or "frameshift". Alternatively, "lof" may by used collectively for loss-of-function classes ("non + splice + frameshift").
Whichever input format is chosen, summary statistics can be produced for "lof", "prot" (protein-altering = mis + lof), and "all".  "prot" and "all" are not valid input classes. In-frame insertions/deletions are currently not eevaluated within the statistical framework.

```{r, include=F}
geneIdentifiers <- names(dplyr::select(denovolyzeR:::pDNM,-class,-value))
```

A variety of gene identifiers may be used.  Valid identifiers recognized by the software include: `r geneIdentifiers`.  The default option ("geneName") specifies gene symbols by default, more specifically, these correspond to the "external\_gene\_name" provided by the Ensembl genome browser {25352552}. ensgID and enstID refer to Ensembl gene and transcript identifiers. hgncID and hgncSymbol refer to HUGO Gene Nomenclature Committee ID numbers and symbols. Within the `R` environment, the `BiomaRt` package from Bioconductor provides tools to convert gene identifiers {19617889}.

3) Compare the observed burden of *de novo* variation to expectation.

The `denovolyzeByClass` function will perform the required analysis.  The function has three required arguments:  

-  `genes`: a vector of gene identifiers, for genes that contain *de novo* variants
-  `classes`: a vector of variant consequences (corresponding to the gene list)
-  `nsamples`: the total number of samples  analyzed (including samples without *de novo* variants).  For the example data, 1078 individuals were sequenced.

```{r}
denovolyzeByClass(genes=autismDeNovos$gene,
                  classes=autismDeNovos$class,
                  nsamples=1078)
```
For each variant class, this function returns the observed number of variants, the expected number of variants, enrichment ($= observed/expected$), and the p value (obtained from a Poisson test).

The output can be customized using the "includeClasses" argument, either to display only a subset of variant classes of interest
```{r}
denovolyzeByClass(genes=autismDeNovos$gene,
                  classes=autismDeNovos$class,
                  nsamples=1078,
                  includeClasses=c("mis","lof"))
```

or to display increased granularity.  By default, nonsense, frameshift & splice variants are analyzed in combination as "lof", but may be analyzed separately.
```{r}
denovolyzeByClass(genes=autismDeNovos$gene,
                  classes=autismDeNovos$class,
                  nsamples=1078,
                  includeClasses=c("frameshift","non","splice","lof"))
```


Further information on function options, and help generally, is available using the help function.
```{r, eval=FALSE}
help(denovolyzeByClass)
```


#BASIC PROTOCOL 2
##Assessing the number of genes with multiple *de novo* variants
The occurrence of multiple *de novo* events in a single gene, in a cohort of individuals with a common phenotype, may implicate that gene in the pathogenesis of the condition under study.  Before evaluating single genes, it is instructive to assess the total number of genes harboring multiple *de novo* variants.  Here, the number of genes containing multiple *de novo* variants is compared with an empirical distribution derived by permutation.

###Materials
As for protocol 1


1) Ensure the **denovolyzeR** library and data for analysis are loaded.
```{r}
library(denovolyzeR)
```

2) The `denovolyzeMultiHits` function will perform the required analysis. 
The same three arguments are required as for BASIC PROTOCOL 1: `genes` (vector of genes containing *de novo* variants), `classes` (a vector of variant consequences) and `nsamples` (number of samples).  In addition, `nperms` determines the number of permutations run (defaults to 100).

The function asks the question "given `nVars` variants in a set of genes, how many genes are expected to contain more than one variant?"
```{r}
denovolyzeMultiHits(genes=autismDeNovos$gene,
                    classes=autismDeNovos$class,
                    nsamples=1078,
                    nperms=100)
```
For each variant class, the function returns the observed number of genes containing multiple *de novo* variants in the user data provided ("obs"), the average number of genes containing multiple hits across `nperms` permutations ("expMean"), the maximum number of genes containing multiple hits in any permutation ("expMax"), and an empirical p value ("pValue").  In this case some of the p values are returned as 0, indicating $< 1/nperms$ (in this case <0.01).  We can obtain a better estimate by increasing the number of permutations: 
```{r}
denovolyzeMultiHits(genes=autismDeNovos$gene,
                    classes=autismDeNovos$class,
                    nsamples=1078,
                    nperms=5000,
                    includeClasses="prot")
```
Note that the exact numbers may change slightly between runs of `denovolyzeMultiHits` due to stochastic changes in the permutations. These stochastic fluctuations are likely to be small for large numbers of permutations.
Finally, the function reports the total number of *de novo* variants of each class, which is the number used as input to the permutation ("nVars").

3) This function can be run in two modes. The expected number of genes containing >1 hit is obtained by permutation: given `nVars` *de novo* variants, how many genes contain >1 variant? There are two options for selecting `nVars`. By default, this number is derived from the input data - in other words, the total number of lof variants that are permuted across the defined gene list is the total number of lof variants in the input data.  An alternative approach uses the expected number of lof variants in the gene list, as determined by the model.  
In the example above autismDeNovos contains
`r sum(denovolyzeR::autismDeNovos$class %in% c("frameshift","non","splice"))`
lof variants, so by default this is the number used in the permutation:
```{r}
sum(autismDeNovos$class %in% c("frameshift","non","splice"))
denovolyzeMultiHits(genes=autismDeNovos$gene,
                    classes=autismDeNovos$class,
                    nsamples=1078,
                    includeClasses="lof")
```
The expected number of *de novo* variants is controlled by the `nVars` argument, whose default value is "actual".  This is a conservative approach, addressing the question: “given the number of variants in our dataset, do we see more genes with >1 variant than expected?”  An alternative approach simply asks whether there are more genes with >1 variant than our *de novo* model predicts. This is accessed by setting `nVars="expected"`.
```{r}
denovolyzeMultiHits(genes=autismDeNovos$gene,
                    classes=autismDeNovos$class,
                    nsamples=1078,
                    includeClasses="lof",
                    nVars="expected")
```


#BASIC PROTOCOL 3
##Assessing the frequency of *de novo* variants in individual genes
In the previous protocol, we assessed whether there were more genes containing multiple *de novo* variants than expected by chance.  In the example data, we noted five genes with multiple loss-of-function hits. In this next protocol, we will determine whether any individual genes carry an excess of *de novo* variants, using the `denovolyzeByGene` function.

###Materials
As for protocol 1

1) Ensure the **denovolyzeR** library and data for analysis are loaded.
```{r}
library(denovolyzeR)
```

2) Call the `denovolyzeByGene` function. The same three arguments are required as for the previous protocols: `genes` (vector of names of genes containing *de novo* variants), `classes` (a vector of variant consequences) and `nsamples` (number of samples).
This function will return one row per gene, ordered according the significance of any enrichment in *de novo* variants.  Given the size of the data, we will only view the first few lines here, using the `head` function.

```{r}
head(
  denovolyzeByGene(genes=autismDeNovos$gene,
                   classes=autismDeNovos$class,
                   nsamples=1078)
  )
```

**denovolyzeR** will output one line for every gene that contains at least one variant in  the input data. In order to view only genes with multiple hits, we can use the subset function to select genes with more than one observed protein-altering variant:
```{r, message=FALSE}
library(dplyr)
denovolyzeByGene(genes=autismDeNovos$gene,
                   classes=autismDeNovos$class,
                   nsamples=1078) %>%
  subset(prot.obs>1)
```
In this example we have used the pipe notation "%>%" to pass the output of **denovolyzeR** to the `subset` function. This is available as part of the `dplyr` package, which is required for denovoloyzeR installation.
 

```{r, include=FALSE}
nTests=length(unique(denovolyzeR:::pDNM$geneName))
bonferroni=signif(0.05/(2*length(unique(denovolyzeR:::pDNM$geneName))),2)
mySciFormat <- function(x){
  # x = number in scientific notation
  x <- unlist(strsplit(as.character(x),"e")) 
  return(paste("$",x[1],"\\!\\times\\!10^{",x[2],"}$",sep=""))
}
```

The p-values returned are not corrected for multiple testing. These default options apply two tests ("lof" and "prot") across `r nTests` genes, so a Bonferroni corrected p-value threshold at α = 0.05 would be `r mySciFormat(bonferroni)` ($0.05 / 2 * `r bonferroni`$).

By default this function compares the number of lof variants against expectation for each gene, and then the total number of protein-altering variants (lof + missense). It can also be configured to return other classes if relevant, using the `includeClasses` argument.
```{r}
head(
  denovolyzeByGene(genes=autismDeNovos$gene,
                   classes=autismDeNovos$class,
                   nsamples=1078,
                   includeClasses="syn")
  )
```




#BASIC PROTOCOL 4
##Assessing a pre-specified gene set
This protocol assesses whether a pre-specified set of genes collectively shows an enrichment of *de novo* variants. Note that any of the previous analyzes can be restricted to a pre-specified geneset in the same way, using the `includeGenes` argument. This may be appropriate if a smaller panel of genes have been sequenced (rather than whole exome sequencing), or to explore biologically relevant genesets, e.g. defined by gene ontology, or expression profile.

###Materials
As for protocol 1

1) Ensure the **denovolyzeR** library and data for analysis are loaded.
```{r, message=FALSE, warning=FALSE}
library(denovolyzeR)
```

2) Define a geneset. This should be a vector of genes, which may be entered by hand, or read from file using `read.table` or equivalent.  In this example, we use an example geneset included with the **denovolyzeR** package, a list of `r nrow(denovolyzeR::fmrpGenes)` genes that interact with the fragile X mental retardation protein (FMRP), taken from {21784246}.
```{r}
nrow(fmrpGenes);head(fmrpGenes)
```

3) Evaluate the frequency of *de novo* events in our pre-specified genelist, using the `denovolyzeByClass` function.  Specify the genelist using the `includeGenes` argument, which defaults to "all", but accepts a vector of genes.
```{r}
denovolyzeByClass(genes=autismDeNovos$gene,
                  classes=autismDeNovos$class,
                  nsamples=1078,
                  includeGenes=fmrpGenes$geneName)
```
In this example we see a highly significant enrichment of *de novo* lof variants in genes that interact with FMRP in our cohort of autism cases. Care should be taken to ensure that the same gene identifiers are used throughout the analysis. For example, if the list of genes containing *de novo* variants includes KMT2D (previously known as MLL2) but the geneset uses the old symbol MLL2, these will not be matched.  The function will give a warning if gene identifiers are used that are not found in the internal mutation probability tables.  

For many genes, the Ensembl gene name and HGNC symbol will be identical, but in some instances they differ (e.g. where there is no HGNC identifier, and Ensembl uses a symbol from another source). Note that we receive a warning if we pass a list of genes described using Ensembl gene symbols (the demonstration data), but tell the software to match to HGNC symbols.
```{r}
denovolyzeByClass(genes=autismDeNovos$gene,
                  classes=autismDeNovos$class,
                  nsamples=1078,
                  geneId="hgncSymbol")
```

Similarly, we will get a warning if "includeGenes" contains non-matching identifiers
```{r}
denovolyzeByClass(genes=autismDeNovos$gene,
                  classes=autismDeNovos$class,
                  nsamples=1078,
                  includeGenes=fmrpGenes$enstID)
```


#ALTERNATIVE PROTOCOL 1
`denovolyzeByClass` and `denovolyzeByGene` are convenience functions, that provide sensible default arguments to the core underlying `denovolyze` function.

The following command
```{r, eval=FALSE}
denovolyzeByClass(genes=autismDeNovos$gene,
                  classes=autismDeNovos$class,
                  nsamples=1078)
```
is identical to
```{r, eval=FALSE}
denovolyze(genes=autismDeNovos$gene,
           classes=autismDeNovos$class,
           nsamples=1078,
           groupBy="class",
           includeClasses=c("syn","mis","lof","prot","all"),
           includeGenes="all"
           )
```

Similarly,
```{r, eval=FALSE}
denovolyzeByGene(genes=autismDeNovos$gene,
                 classes=autismDeNovos$class,
                 nsamples=1078)
```
is identical to
```{r, eval=FALSE}
denovolyze(genes=autismDeNovos$gene,
           classes=autismDeNovos$class,
           nsamples=1078,
           groupBy="gene",
           includeClasses=c("lof","prot"),
           includeGenes="all"
           )
```


#SUPPORT PROTOCOL 1
##Getting help
Help on any of the functions described is available using the standard `R` help functions, e.g. `help(denovolyze)` or `?denovolyze`.  Additional details are also available in the package vignette, accessed using `vignette("denovolyzeR_intro")`.

#SUPPORT PROTOCOL 2
##Viewing the mutational probability tables
Users may want to view or export the probability tables that underpin these analyzes.  These are best accessed using the `viewProbabilityTable` function.

```{r viewProbabilityTable}
probabilityTable <- viewProbabilityTable()
nrow(probabilityTable); head(probabilityTable)
```

This may be useful, for example, to verify that the input gene list contains the correct identifiers
```{r,collapse=FALSE}
#Count the number of input gene names
length(autismDeNovos$gene)
#Count how many are in the "geneName" column of the probability table:
sum(autismDeNovos$gene %in% probabilityTable$geneName)
#Count how many are in the "hgncSymbol" column of the probability table:
sum(autismDeNovos$gene %in% probabilityTable$hgncSymbol)
#Count how many are in the "enstID" column of the probability table:
sum(autismDeNovos$gene %in% probabilityTable$enstID)
```

#SUPPORT PROTOCOL 3
##Using an alternative mutational probability table
**denovolyzeR** relies on a pre-computed tabulation of the probability of *de novo* variation arising in each gene, as described in the Introduction and Background Information.  The default probability table was generated by calculating the probability of *de novo* events for every base of the canonical Gencode transcripts, as defined in Gencode v19.  It is beyond the scope of this protocol to describe methods to compute these tables, but **denovolyzeR** does allow for the import of alternative tables, if required.
For example, the original paper describing this analytical framework {25086666} calculated mutational probabilities based on  RefSeq transcript definitions, whereas **denovolyzeR** now uses Gencode definitions.  Tables may also be computed to include other functional consequences (e.g. damaging missense variants, as determined by *in silico* SNV consequence prediction algorithms).

###Materials
An alternative probability table. Examples are available to download from http://denovolyzer.org/

1) Locate and load the chosen probability table.  For this example, we will use "probTable_Samocha2014.rda" downloaded from the above link, which is now located in our Downloads folder:

```{r, include=FALSE}
pathToProbabilityTable="alternativeProbabilityTables/"
```

```{r altProbabilityTable}
#pathToProbabilityTable="~/Downloads" #replace this with the path on your local system
load(file.path(pathToProbabilityTable,"probTable_Samocha2014.rda"))
head(probTable_Samocha2014)
```
This table has two sets of gene identifiers: "refseqID"", and gene symbols ("geneID").

2) Run chosen analysis, specifying the chosen probability table using the "probTable" argument, and the appropriate gene identifier.
```{r}
denovolyzeByClass(genes=autismDeNovos$gene,
                  classes=autismDeNovos$class,
                  nsamples=1078,
                  probTable=probTable_Samocha2014,
                  geneId="geneID")
```
In this instance there is a warning that not all of the input identifiers are recognized. This is because there is not a one-to-one mapping between the identifiers associated with RefSeq and Gencode transcripts.

_____

##COMMENTARY

####Background Information  
The mutational model is described in full detail in {25086666}. Briefly, it is based on a determination of the probability of each base in the coding sequence of the human genome mutating to each of the other possible bases. The predicted impact of these changes is aggregated across the gene to establish the probability of specific types of mutation (synonymous, missense, etc).

Previous work established that the mutability of a base is sufficiently modeled by accounting for the local sequence context of one nucleotide on either side of the base of interest {9683596;17357078}. We analyzed human variation and trinucleotide context using intergenic single nucleotide polymorphisms (SNPs) from the 1000 Genomes project to create a mutation rate table, which provides the relative probabilities of each possible trinucleotide (XY^1^Z) to trinucleotide (XY^2^Z) change.

We then consider each base in the coding sequence and use the mutation rate table to determine its probability of mutating to the other bases. We predict the impact of the mutation on the protein product and aggregate the probabilities by mutation type across each gene. These per-gene probabilities are then adjusted for how well sequenced the gene was and regional divergence between humans and macaques.

Given that the number of *de novo* variants per trio follows a Poisson distribution {22495311}, we use the Poisson distribution to evaluate the excesses of *de novo* events. As an example, to determine if a particular gene has more *de novo* loss-of-function variants than expected, we multiply that gene’s probability of a loss-of-function mutation by the number of trios and by 2 (for the number of chromosomes) in the study. This gives the expected number of *de novo* loss-of-function variants (denoted as lambda, $\lambda$) with which the observed number is compared. Specifically, the `ppois` command in `R` is used. The three *de novo* loss-of-function variants seen in DYRK1A are used in the example below. With `denovolyze`, we get the following result:
```{r}
denovolyzeByGene(genes=autismDeNovos$gene,
                 classes=autismDeNovos$class,
                 nsamples=1078,
                 includeGenes="DYRK1A")
```

We can reproduce this with the `ppois` function.  Note that by default, `ppois(q,lambda)` will return the probability of observing $\leq q$ events for a given $\lambda$.  We are interested in computing  $p(obs \geq q)$.  `ppois(q,lambda,lower.tail=FALSE)` gives us $p(obs > q)$, and therefore we must use `ppois(q-1,lambda,lower.tail=FALSE)` to obtain $p(obs \geq q)$.

```{r, collapse=FALSE}
n_lof_dyrk1a <- 3
probabilityTable[probabilityTable$geneName=="DYRK1A","lof"]
prob_lof_dyrk1a <- 2.528297e-06
n_trios <- 1078
ppois(q=n_lof_dyrk1a-1, #observed - 1
      lambda=prob_lof_dyrk1a*n_trios*2, #expected
      lower.tail=FALSE)
```

**Control subjects**
Since this analytic framework compares *de novo* events to a model-derived expectation, there is no direct comparison of cases with controls.  As described above, direct case-control comparison is not statistically powerful in this context.  It is nonetheless recommended to include a control arm in analyses, for example by repeating the analyses described above on a cohort of controls which have been subjected to the same sequencing and *de novo* identification pipeline.  While cases and controls are not directly compared, it is valuable to confirm that the burden of variants in the control population does not deviate substantially from model-derived expectations, in order to validate the model for the specific sequencing approaches used in each study.

####Critical Parameters   
The number of samples (`nsamples`) should be the total number of samples in the study, not just those that carry *de novo* variants.  
The analysis described in BASIC PROTOCOL 2 is highly sensitive to the choice of argument passed to `nVars`. Full details are provided in step 3 of that protocol.


####Troubleshooting  
Leave blank

####Anticipated Results  
Leave blank

####Time Considerations  
These analyzes are not especially computationally intensive, and will run on a desktop or laptop computer in seconds.  The `denovolyzeMultiHits` function uses permutation, and computation time increases linearly with the number of permutations.  Elapsed times (in seconds) to run the three principal functions on *de novo* variants from 1078 samples, using default settings, on a MacBook Air (1.7GHz i7, 8Gb RAM) are as follows:
```{r, collapse=FALSE}
system.time(denovolyzeByClass(genes=autismDeNovos$gene,classes=autismDeNovos$class,nsamples=1078))["elapsed"]
system.time(denovolyzeMultiHits(genes=autismDeNovos$gene,classes=autismDeNovos$class,nsamples=1078,nperms=1000))["elapsed"]
system.time(denovolyzeByGene(genes=autismDeNovos$gene,classes=autismDeNovos$class,nsamples=1078))["elapsed"]
```


##ACKNOWLEDGEMENT
JSW is supported by the NIHR Biomedical Research Unit in Cardiovascular Disease at Royal Brompton
& Harefield NHS Foundation Trust and Imperial College London, Academy of Medical Sciences, Wellcome Trust, British Heart Foundation Arthritis Research UK, Fondation Leducq, and  an HICF award from Wellcome Trust and Department of Health.  
JGH is supported by the John S. LaDue Fellowship at Harvard Medical School.

##LITERATURE CITED

##KEY REFERENCE (optional)
{25086666}

##INTERNET RESOURCES (optional)

http://www.r-project.org/  
http://denovolyzer.org/  
http://jamesware.github.io/denovolyzeR/  

##FIGURE LEGENDS
na

##TABLES
na